File: leetcode/1326.cc
#include <vector>
#include <utility>
#include <algorithm>


using namespace std;


class Solution {
  public:
    int minTaps(int n, vector<int>& ranges) {
        vector<pair<int, int>> inters;
        for (int i = 0; i <= n; ++i)
            inters.emplace_back(i-ranges[i], i+ranges[i]);
        sort(inters.begin(), inters.end());

        int cur = 0;
        int idx = 0;
        int cnt = 0;
        while (cur < n) {
            int next = cur;
            while (inters[idx].first <= cur) {
                next = max(next, inters[idx].second);
                ++idx;
            }
            if (next == cur)
                return -1;
            cur = next;
            ++cnt;
        }
        return cnt;
    }
};

File: leetcode/1326.py
class Solution:
    def minTaps(self, n: int, ranges: List[int]) -> int:
        assert n > 0
        inter = []
        for i, r in enumerate(ranges):
            inter.append([i - r, i + r])
        inter.sort()

        cnt = 0
        farthest = 0
        i = 0
        while farthest < n:
            next_reach = farthest
            while i <= n and inter[i][0] <= farthest:
                next_reach = max(next_reach, inter[i][1])
                i += 1
            if next_reach == farthest:
                return -1
            cnt += 1
            farthest = next_reach
        return cnt
